libs/shared/providers/src/wagmi/components/TransactionButton.tsx

```diff
@@ -0,0 +1,218 @@
+import { useEffect, useState } from 'react';
+
+import { Button } from '@mui/material';
+import { NotificationSnack, SeverityIcon } from '@origin/shared/components';
+import { isNilOrEmpty, isUserRejected } from '@origin/shared/utils';
+import { useIntl } from 'react-intl';
+import {
+  useAccount,
+  useContractWrite,
+  usePrepareContractWrite,
+  useWaitForTransaction,
+} from 'wagmi';
+
+import {
+  SwapNotification,
+  useDeleteActivity,
+  usePushActivity,
+  useUpdateActivity,
+} from '../../activities';
+import { usePushNotification } from '../../notifications';
+
+import type { ButtonProps } from '@mui/material';
+import type { Contract } from '@origin/shared/contracts';
+import type { TransactionReceipt } from 'viem';
+
+import type { Activity } from '../../activities';
+
+export type TransactionButtonProps = {
+  contract: Contract;
+  functionName: string;
+  args?: unknown[];
+  value?: bigint;
+  label?: string;
+  onSuccess?: (txReceipt: TransactionReceipt) => void;
+  onError?: (error: Error) => void;
+  onUserReject?: () => void;
+  disableActivity?: boolean;
+  disableNotification?: boolean;
+} & Omit<ButtonProps, 'onClick'>;
+
+export const TransactionButton = ({
+  contract,
+  functionName,
+  args,
+  value,
+  label,
+  onSuccess,
+  onError,
+  onUserReject,
+  disabled,
+  disableActivity,
+  disableNotification,
+  ...rest
+}: TransactionButtonProps) => {
+  const intl = useIntl();
+  const { isConnected } = useAccount();
+  const [activity, setActivity] = useState<Activity>(null);
+  const pushNotification = usePushNotification();
+  const pushActivity = usePushActivity();
+  const updateActivity = useUpdateActivity();
+  const deleteActivity = useDeleteActivity();
+  const { config } = usePrepareContractWrite({
+    address: contract.address,
+    abi: contract.abi,
+    functionName,
+    args,
+    value,
+    enabled: isConnected && !!contract?.address,
+  });
+  const {
+    write,
+    data: writeData,
+    error: writeError,
+    isLoading: isWriteLoading,
+  } = useContractWrite(config);
+  const {
+    data: txData,
+    isLoading: isTxLoading,
+    isSuccess: isTxSuccess,
+  } = useWaitForTransaction({ hash: writeData?.hash });
+
+  useEffect(() => {
+    if (!isNilOrEmpty(txData) && !isTxLoading && isTxSuccess) {
+      if (onSuccess) {
+        onSuccess(txData);
+      }
+      if (!disableActivity) {
+        updateActivity({
+          ...activity,
+          status: 'success',
+          txReceipt: txData,
+        });
+      }
+      if (!disableNotification) {
+        pushNotification({
+          content: (
+            <SwapNotification
+              {...activity}
+              status="success"
+              txReceipt={txData}
+            />
+          ),
+        });
+      }
+    }
+  }, [
+    activity,
+    disableActivity,
+    disableNotification,
+    isTxLoading,
+    isTxSuccess,
+    onSuccess,
+    pushNotification,
+    txData,
+    updateActivity,
+  ]);
```

Comment:
Since I've been working on subsquid for so long, I can say I don't miss hook dependencies.

Reply:
eslint to the rescue!
![image](https://github.com/OriginProtocol/origin-defi/assets/13216617/730c5c42-0420-483c-97c6-7d3113c9bb4a)